hudi-spark-datasource/hudi-spark3-common/src/main/scala/org/apache/spark/sql/execution/datasources/Spark3ParquetSchemaEvolutionUtils.scala

```diff
@@ -0,0 +1,194 @@
+        case AlwaysTrue =>
+          AlwaysTrue
+        case AlwaysFalse =>
+          AlwaysFalse
+        case _ =>
+          AlwaysTrue
+      }
+    }
+  }
+
+  protected var typeChangeInfos: java.util.Map[Integer, Pair[DataType, DataType]] = null
+
+  def getHadoopConfClone(footerFileMetaData: FileMetaData, enableVectorizedReader: Boolean): Configuration = {
+    // Clone new conf
+    val hadoopAttemptConf = new Configuration(sharedConf)
+    typeChangeInfos = if (shouldUseInternalSchema) {
+      val mergedInternalSchema = new InternalSchemaMerger(fileSchema, querySchemaOption.get(), true, true).mergeSchema()
+      val mergedSchema = SparkInternalSchemaConverter.constructSparkSchemaFromInternalSchema(mergedInternalSchema)
+
+      hadoopAttemptConf.set(ParquetReadSupport.SPARK_ROW_REQUESTED_SCHEMA, mergedSchema.json)
+
+      SparkInternalSchemaConverter.collectTypeChangedCols(querySchemaOption.get(), mergedInternalSchema)
+    } else {
+      val (implicitTypeChangeInfo, sparkRequestSchema) = HoodieParquetFileFormatHelper.buildImplicitSchemaChangeInfo(hadoopAttemptConf, footerFileMetaData, requiredSchema)
+      if (!implicitTypeChangeInfo.isEmpty) {
+        shouldUseInternalSchema = true
+        hadoopAttemptConf.set(ParquetReadSupport.SPARK_ROW_REQUESTED_SCHEMA, sparkRequestSchema.json)
+      }
+      implicitTypeChangeInfo
+    }
+
+    if (enableVectorizedReader && shouldUseInternalSchema &&
+      !typeChangeInfos.values().forall(_.getLeft.isInstanceOf[AtomicType])) {
+      throw new IllegalArgumentException(
+        "Nested types with type changes(implicit or explicit) cannot be read in vectorized mode. " +
+          "To workaround this issue, set spark.sql.parquet.enableVectorizedReader=false.")
+    }
+
+    hadoopAttemptConf
+  }
+
+  def generateUnsafeProjection(fullSchema: Seq[AttributeReference], timeZoneId: Option[String]): UnsafeProjection = {
+
+    if (typeChangeInfos.isEmpty) {
+      GenerateUnsafeProjection.generate(fullSchema, fullSchema)
+    } else {
+      // find type changed.
+      val newSchema = new StructType(requiredSchema.fields.zipWithIndex.map { case (f, i) =>
+        if (typeChangeInfos.containsKey(i)) {
+          StructField(f.name, typeChangeInfos.get(i).getRight, f.nullable, f.metadata)
+        } else f
+      })
+      val newFullSchema = toAttributes(newSchema) ++ toAttributes(partitionSchema)
+      val castSchema = newFullSchema.zipWithIndex.map { case (attr, i) =>
+        if (typeChangeInfos.containsKey(i)) {
+          val srcType = typeChangeInfos.get(i).getRight
+          val dstType = typeChangeInfos.get(i).getLeft
+          val needTimeZone = Cast.needsTimeZone(srcType, dstType)
+          Cast(attr, dstType, if (needTimeZone) timeZoneId else None)
+        } else attr
+      }
+      GenerateUnsafeProjection.generate(castSchema, newFullSchema)
+    }
+  }
+
+  protected def toAttributes(schema: StructType): Seq[AttributeReference]
```

Comment:
There is `HoodieSchemaUtils.toAttributes`.  Can that be reused, so there is no need to create this function here?

Reply:
yes that can be reused